aptos-move/framework/aptos-framework/sources/coin_supply.move

```diff
@@ -0,0 +1,88 @@
+/// This module contains logic for upgradable and parallelizable coin supply
+/// tracking in Move. It is a seprate module in order to hide aggregator from
+/// the coin.
+///
+///  +--------+     +---------------+     +-----------------------+
+///  |  coin  | --> |  coin_supply  | --> |  optional_aggregator  |
+///  +--------+     +---------------+     +-----------------------+
+///                                         coin doesn't know
+///                                         this exists!
+///
+/// This is not u128, and is intnded to be used from `coin` module only. Thus,
+/// only `coin` is allowed to create or upgrade the supply struct.
+///
+/// `Supply` is stored within `CoinInfo` and is updated when coins are minted
+/// or burnt. In addition, one can get the coin supply in existence by calling
+/// `coin::supply<CoinType>()` that uses API from this module under the hood.
+module aptos_framework::coin_supply {
+    use aptos_framework::optional_aggregator::{Self, OptionalAggregator};
+    use aptos_framework::system_addresses;
+    use aptos_std::type_info;
+
+    friend aptos_framework::coin;
+
+    #[test_only]
+    friend aptos_framework::coin_supply_tests;
+
+    /// Maximum possible coin supply.
+    const MAX_U128: u128 = 340282366920938463463374607431768211455;
+
+    /// Tracks supply of coins of type `CoinType` in the system.
+    struct Supply<phantom CoinType> has store {
+        inner: OptionalAggregator,
+    }
+
+    /// Creates a new supply tracker for `CoinType`.
+    public(friend) fun new<CoinType>(): Supply<CoinType> {
+        let type_info = type_info::type_of<CoinType>();
+        let addr = type_info::account_address(&type_info);
+        new_from_address(addr)
+    }
+
+    fun new_from_address<CoinType>(addr: address): Supply<CoinType> {
+        if (system_addresses::is_aptos_framework_address(addr)) {
```

Comment:
I don't 100% like having this logic here as it makes coin_supply, something that seems generic, aware of the difference between AptosCoin and other coins. I also don't think you need this entire separate coin_supply module just to be a wrapper around optional aggregator. Why not merge all of these logic into the coin module and have a separate initialize function that can only be called by aptos_coin module that allows initialize the optional aggregator with parallelizable = true?